File: iac/pulumi/storage.ts
/**
 * GCS buckets for static frontend hosting (app + docs).
 *
 * Deployment:
 *   yarn --cwd app build
 *   gsutil -m rsync -r -d app/dist gs://app-classifier-tabiya-tech
 *
 *   yarn --cwd docs build
 *   gsutil -m rsync -r -d docs/dist gs://docs-classifier-tabiya-tech
 *
 * CDN invalidation (if needed):
 *   gcloud compute url-maps invalidate-cdn-cache LOAD_BALANCER_NAME --path "/*"
 */

import * as gcp from "@pulumi/gcp";

interface Args {
  project: string;
  region: string;
}

function createStaticBucket(name: string, project: string) {
  const bucket = new gcp.storage.Bucket(name, {
    project,
    name,
    location: "US",
    uniformBucketLevelAccess: true,
    website: {
      mainPageSuffix: "index.html",
      notFoundPage: "index.html",  // SPA fallback
    },
    cors: [
      {
        origins: ["*"],
        methods: ["GET", "HEAD", "OPTIONS"],
        responseHeaders: ["Content-Type"],
        maxAgeSeconds: 3600,
      },
    ],
  });

  // Allow public read
  new gcp.storage.BucketIAMMember(`${name}-public`, {
    bucket: bucket.name,
    role: "roles/storage.objectViewer",
    member: "allUsers",
  });

  // Backend bucket for Cloud CDN
  const backendBucket = new gcp.compute.BackendBucket(`${name}-backend`, {
    project,
    name: name.replace(/\./g, "-"),
    bucketName: bucket.name,
    enableCdn: true,
    cdnPolicy: {
      cacheMode: "CACHE_ALL_STATIC",
      defaultTtl: 3600,
      maxTtl: 86400,
    },
  });

  return { bucket, backendBucket };
}

export function createFrontendBuckets({ project, region }: Args) {
  const app = createStaticBucket("app-classifier-tabiya-tech", project);
  const docs = createStaticBucket("docs-classifier-tabiya-tech", project);

  // ── URL maps and HTTPS load balancers ────────────────────────────────────

  const appUrlMap = new gcp.compute.URLMap("app-url-map", {
    project,
    name: "app-classifier-tabiya-tech-lb",
    defaultService: app.backendBucket.selfLink,
  });

  const docsUrlMap = new gcp.compute.URLMap("docs-url-map", {
    project,
    name: "docs-classifier-tabiya-tech-lb",
    defaultService: docs.backendBucket.selfLink,
  });

  // GCP-managed SSL certificates
  const appCert = new gcp.compute.ManagedSslCertificate("app-cert", {
    project,
    name: "app-classifier-tabiya-tech-cert",
    managed: { domains: ["app.classifier.tabiya.tech"] },
  });

  const docsCert = new gcp.compute.ManagedSslCertificate("docs-cert", {
    project,
    name: "docs-classifier-tabiya-tech-cert",
    managed: { domains: ["docs.classifier.tabiya.tech"] },
  });

  // HTTPS proxies
  const appProxy = new gcp.compute.TargetHttpsProxy("app-proxy", {
    project,
    name: "app-classifier-tabiya-tech-proxy",
    urlMap: appUrlMap.selfLink,
    sslCertificates: [appCert.selfLink],
  });

  const docsProxy = new gcp.compute.TargetHttpsProxy("docs-proxy", {
    project,
    name: "docs-classifier-tabiya-tech-proxy",
    urlMap: docsUrlMap.selfLink,
    sslCertificates: [docsCert.selfLink],
  });

  // Global forwarding rules (port 443)
  new gcp.compute.GlobalForwardingRule("app-forwarding-rule", {
    project,
    name: "app-classifier-tabiya-tech-fw",
    target: appProxy.selfLink,
    portRange: "443",
    loadBalancingScheme: "EXTERNAL_MANAGED",
  });

  new gcp.compute.GlobalForwardingRule("docs-forwarding-rule", {
    project,
    name: "docs-classifier-tabiya-tech-fw",
    target: docsProxy.selfLink,
    portRange: "443",
    loadBalancingScheme: "EXTERNAL_MANAGED",
  });

  return { appBucket: app.bucket, docsBucket: docs.bucket };
}

File: iac/pulumi/storage.py
"""GCS buckets for static frontend hosting (app + docs).

Deployment:
  yarn --cwd app build
  gsutil -m rsync -r -d app/dist gs://app-classifier-tabiya-tech

  yarn --cwd docs build
  gsutil -m rsync -r -d docs/dist gs://docs-classifier-tabiya-tech

CDN invalidation (if needed):
  gcloud compute url-maps invalidate-cdn-cache LOAD_BALANCER_NAME --path "/*"
"""

import pulumi_gcp as gcp


def _create_static_bucket(name: str, project: str):
    bucket = gcp.storage.Bucket(
        name,
        project=project,
        name=name,
        location="US",
        uniform_bucket_level_access=True,
        website=gcp.storage.BucketWebsiteArgs(
            main_page_suffix="index.html",
            not_found_page="index.html",  # SPA fallback
        ),
        cors=[
            gcp.storage.BucketCorArgs(
                origins=["*"],
                methods=["GET", "HEAD", "OPTIONS"],
                response_headers=["Content-Type"],
                max_age_seconds=3600,
            )
        ],
    )

    # Allow public read
    gcp.storage.BucketIAMMember(
        f"{name}-public",
        bucket=bucket.name,
        role="roles/storage.objectViewer",
        member="allUsers",
    )

    # Backend bucket for Cloud CDN
    backend_bucket = gcp.compute.BackendBucket(
        f"{name}-backend",
        project=project,
        name=name.replace(".", "-"),
        bucket_name=bucket.name,
        enable_cdn=True,
        cdn_policy=gcp.compute.BackendBucketCdnPolicyArgs(
            cache_mode="CACHE_ALL_STATIC",
            default_ttl=3600,
            max_ttl=86400,
        ),
    )

    return bucket, backend_bucket


def create_frontend_buckets(project: str, region: str):
    app_bucket, app_backend = _create_static_bucket("app-classifier-tabiya-tech", project)
    docs_bucket, docs_backend = _create_static_bucket("docs-classifier-tabiya-tech", project)

    # ── URL maps and HTTPS load balancers ─────────────────────────────────────

    app_url_map = gcp.compute.URLMap(
        "app-url-map",
        project=project,
        name="app-classifier-tabiya-tech-lb",
        default_service=app_backend.self_link,
    )

    docs_url_map = gcp.compute.URLMap(
        "docs-url-map",
        project=project,
        name="docs-classifier-tabiya-tech-lb",
        default_service=docs_backend.self_link,
    )

    # GCP-managed SSL certificates
    app_cert = gcp.compute.ManagedSslCertificate(
        "app-cert",
        project=project,
        name="app-classifier-tabiya-tech-cert",
        managed=gcp.compute.ManagedSslCertificateManagedArgs(
            domains=["app.classifier.tabiya.tech"]
        ),
    )

    docs_cert = gcp.compute.ManagedSslCertificate(
        "docs-cert",
        project=project,
        name="docs-classifier-tabiya-tech-cert",
        managed=gcp.compute.ManagedSslCertificateManagedArgs(
            domains=["docs.classifier.tabiya.tech"]
        ),
    )

    # HTTPS proxies
    app_proxy = gcp.compute.TargetHttpsProxy(
        "app-proxy",
        project=project,
        name="app-classifier-tabiya-tech-proxy",
        url_map=app_url_map.self_link,
        ssl_certificates=[app_cert.self_link],
    )

    docs_proxy = gcp.compute.TargetHttpsProxy(
        "docs-proxy",
        project=project,
        name="docs-classifier-tabiya-tech-proxy",
        url_map=docs_url_map.self_link,
        ssl_certificates=[docs_cert.self_link],
    )

    # Global forwarding rules (port 443)
    gcp.compute.GlobalForwardingRule(
        "app-forwarding-rule",
        project=project,
        name="app-classifier-tabiya-tech-fw",
        target=app_proxy.self_link,
        port_range="443",
        load_balancing_scheme="EXTERNAL_MANAGED",
    )

    gcp.compute.GlobalForwardingRule(
        "docs-forwarding-rule",
        project=project,
        name="docs-classifier-tabiya-tech-fw",
        target=docs_proxy.self_link,
        port_range="443",
        load_balancing_scheme="EXTERNAL_MANAGED",
    )

    return app_bucket, docs_bucket
